### ISSUE TITLE
Vimium on new-tab page

### ISSUE THREAD

**Issue by smblott-github** [philc/vimium#1515]
When creating a new tab with `Ctrl-T`, Vimium is not active and the focus is locked in the address bar.  The whole Vimium UX is broken.

- Does anyone know how to fix this?

Absent a Vimium fix, I've been using [New Tab Redirect](https://chrome.google.com/webstore/detail/new-tab-redirect/icpgjfneehieebagbmdbhnlpiopdcmna?hl=en) to redirect `chrome://newtab` to a blank page with a very-long TTL (on which Vimium is active).  Unfortunately, we can't do the same trick in Vimium.  First, it completely disables `chrome://newtab` (and it wouldn't be possible to have an option to re-enable it).  Second, redirecting to `pages/blank.html` still leaves the focus in the address bar.

As a super lightweight alternative to New Tab Redirect, I implemented this:
- https://chrome.google.com/webstore/detail/new-tab-url/njigpponciklokfkoddampoienefegcl
- https://github.com/smblott-github/chrome-new-tab-url

(which might be of interest to some Vimium users, absent a Vimium solution).

**Comment by nikess** [comment]
Anyone know any way to make it work with Momentum? (https://chrome.google.com/webstore/detail/momentum/laookkfknpbbblfpciffpaejjkokdgca?utm_source=chrome-app-launcher-info-dialog)

**Comment by smblott-github** [comment]
It'll never work with Momentum, I think, because the new-tab URL is:

    chrome-extension://laookkfknpbbblfpciffpaejjkokdgca

... and Chrome doesn't run other extensions on such pages.

**Comment by leoj3n** [comment]
My experience is different than yours.

Setting the new tab page to `pages/blank.html` in Vimium Options and pressing `t` opens a new tab but does not leave focus on address bar. I can immediately press `o`, `shift+h` or any other Vimium shortcut.

Is your plugin really necessary? Don't Chrome's default new tab settings suffice?

**Comment by smblott-github** [comment]
> Setting the new tab page to pages/blank.html in Vimium Options and pressing t opens a new tab but does not leave focus on address bar

@leoj3n. You're right, for tabs opened with `t`.  My extension then gives the same effect for tabs opened with `Ctrl-T`.

**Comment by leoj3n** [comment]
Ohhh got ya.

**Comment by wadkar** [comment]
I think I might need to open a new issue - but it appears that only I am facing it.
If I put `pages/blank.html` in the vimium settings, it opens a search result with `pages/blank.html` as query. I need to use `chrome-extension://aaklmmoijnlnbeogmilbplamkggababp/pages/blank.html` to actually get the `blank.html` page from vimium's extension source.
Any suggestions/hints on looking for solution/bugfix?

**Comment by mrmr1993** [comment]
The [docs for `chrome.tabs.create`](https://developer.chrome.com/extensions/tabs#method-create) say (of the `url` parameter, which we pass as the contents of this setting):

> Relative URLs will be relative to the current page within the extension.

If this isn't working properly, it sounds like a Chrome bug. Have you tried restarting the browser/reloading Vimium, to see if that makes a difference?

**Comment by smblott-github** [comment]
@mrmr1993.  We can wait to see what @wadkar reports.  But we could also just convert `pages/blank.html` to an absolute URL, and not worry about it.

**Comment by mrmr1993** [comment]
I don't think this is something we need to consider unless this is an issue for multiple people. But it would be tidier UX to just resolve relative urls manually (like chrome says it should), if this is an issue. A `chrome-extension` url looks kinda scary for people who aren't familiar with them.

**Comment by wadkar** [comment]
Test Report
Methodology: Open chrome, load the extension, ensure appropriate settings, run tests, remove the extension
 - Test 1 with Vimium from chrome store (v1.49): Happy Path 1
1. Chrome home page set to about:blank and vimium new tab URL set to `pages/blank.html`
2. Open google.com in normal window, press `t`, a new tab with `chrome-extension://.../pages/blank.html` opens
3. Open an incognito window, press `t`, a new tab with `chrome-extension://.../pages/blank.html` opens _in the non-incognito window_
(Issue #1507 reported , patch submitted PR #1508 and accepted)
 - Test 2 with Vimium from local with PR#1508 applied: Happy Path 2
1. Chrome home page set to about:blank and vimium new tab URL set to `chrome://newtab`
2. Open google.com, press `t`, a new tab with `chrome://newtab` opens
3. Open an incognito window, press `t`, a new tab with `about:blank` opens _in the non-incognito window_
 - Test 3 with Vimium from local with PR #1508 applied: Happy Path 3
1. Chrome home page set to about:blank and vimium new tab URL set to `chrome-extension://.../pages/blank.html`
2. Open google.com, press `t`, a new tab with `chrome-extension://.../pages/blank.html` opens
3. Open an incognito window, press `t`, a new tab with `chrome-extension://.../pages/blank.html` opens _in the non-incognito window_
 - Test 4 with Vimium from local with PR #1508 applied: Failure Path
1. Chrome home page set to about:blank and vimium new tab URL set to `pages/blank.html`
2. Open google.com in normal window, press `t`, a new tab with `pages/blank.html` _does not open_
Instead, I see search result for the query `pages/blank.html`
3. Open an incognito window, press `t`, a new tab with search result for query `pages/blank.html` opens in the incognito window

Any other testing suggestions? My js-foo is a bit rusty, so I didn't attempt full debugging of PR #1508 . If you have any suggestions/hints I will be happy to give it another shot to PR #1508 . Till then, I think we should block PR #1508.

Cheers,
-S

**Comment by wadkar** [comment]
After little bit of digging around, it appears that `createNewUrlInTab` calls [`Utils.convertToUrl`](https://github.com/philc/vimium/blob/master/lib/utils.coffee#L104)
A simple fix is to test for `pages/blank.html` in the `convertToUrl` function as follows:

```coffeescript
  convertToUrl: (string) ->
    string = string.trim()

    # Special-case about:[url], view-source:[url] and the like
    if Utils.hasChromePrefix string
      string
    else if Utils.isUrl string
      Utils.createFullUrl string
    else if string is "pages/blank.html"
      string
    else
      Utils.createSearchUrl string
```
This may or may not be the right way to go. Note that the fix introduces an extension specific string literal `"pages/blank.html"` in the middle of `Utils` library.

If this is acceptable then let me know and I will create another issue/PR with the fix.

Cheers,
-S

**Comment by smblott-github** [comment]
Thanks for detective work, @wadkar.  With the fix you propose, typing `pages/blank.html` into the vomnibar would take you to the blank page.  That seems wrong.

A better alternative would be to find where the new tab setting is used ([here](https://github.com/philc/vimium/blob/master/background_scripts/main.coffee#L248)), and replace it with the full URL directly.  This is the only place where `pages/blank.html` has a special meaning.

**Comment by mrmr1993** [comment]
Changing

```coffee
createTab: (callback) -> openUrlInNewTab { url: Settings.get("newTabUrl") }, callback
```
to
```coffee
createTab: (callback) -> openUrlInNewTab {
  url: Settings.get("newTabUrl")
  rawUrl: true
}, callback
```
and
```coffee
openUrlInNewTab = (request, callback) ->
  chrome.tabs.getSelected null, (tab) ->
    tabConfig =
      url: Utils.convertToUrl request.url
      index: tab.index + 1
      selected: true
      windowId: tab.windowId
    # FIXME(smblott). openUrlInNewTab is being called in two different ways with different arguments.  We
    # should refactor it such that this check on callback isn't necessary.
    callback = (->) unless typeof callback == "function"
    chrome.tabs.create tabConfig, callback
```
to
```coffee
openUrlInNewTab = (request, callback) ->
  chrome.tabs.getSelected null, (tab) ->
    chrome.tabs.create {
      url: if request.rawUrl then request.url else Utils.convertToUrl request.url
      index: tab.index + 1
      selected: true
      windowId: tab.windowId
    # FIXME(smblott). openUrlInNewTab is being called in two different ways with different arguments.  We
    # should refactor it such that this check on callback isn't necessary.
    }, -> callback?()
```
should achieve exactly what we want (and IMO makes `openUrlInNewTab` easier to understand as a `chrome.tabs.create` call). We could also change `openOptionsPageInNewTab` to use `openUrlInNewTab {url: "pages/options.html", rawUrl: true}` too, to avoid any `chrome.tabs.create` duplicated logic. I can't make a PR for this at the moment, unfortunately.

**Comment by wadkar** [comment]
Yep, I agree `Utils` isn't the right way to go. Interestingly, the fix passes Test 2 completely (new tab opens in incognito window) but fails Test 3 by opening `chrome-extension://.../pages/blank.html` in the _non-incognito window_ . I thought about fixing it some other way and arrived a solution quite very much like @mrmr1993 

I think we should create a separate issue/PR for this and link #1507 #1508 against it. I am making a separate branch for this and testing @mrmr1993 's suggestion - stay tuned.

Cheers,
-S

**Comment by wadkar** [comment]
So, I followed through @mrmr1993 's changes and created PR #1532 . It certainly fixes some issues (setting aside incognito mode).

To be honest, I am not sure as to why [Test 1 above](https://github.com/philc/vimium/issues/1515#issuecomment-80079217) should ever pass for incognito windows. I don't suppose chrome allows opening of `chrome-extension://...` URL in an incognito window. Chrome seems to redirect it to non-incognito window **by design**. Perhaps I missed something? Could someone please try out [Test 3 above](https://github.com/philc/vimium/issues/1515#issuecomment-80079217) again?

As such, setting `newTabUrl` option in vimium to `pages/blank.html` should only work for non incognito windows. In other words, providing a `blank.html` with extension will not work for users working in incognito mode. Those who seek consistency throughout chrome and vimium's settings can still apply PR #1532 and set chrome's new tab setting as well as vimium's `newTabUrl` to [some blank page](http://this-page-intentionally-left-blank.org/).

**Comment by gdh1995** [comment]
@leoj3n :
Setting the new tab page to pages/blank.html in Vimium Options and pressing t opens a new tab but does not leave focus on address bar

As I understand it, leoj3n has set Chrome's new tab url to `chrome-extension://***/pages/blank.html`, or use an extension to redirect new tab to such a url.
If so, here are my solution:
I find, Chrome will treat any other new tab pages as `chrome://newtab`. That is, Chrome must focus on its omni bar, no matter what extension the page belongs to.
Then, We have to make a jump:
* add newtab.html to Vimium: 
`<!DOCTYPE HTML><html><head><script type="text/javascript" src="newtab.js"></script></head><body></body></html>`
* add newtab.js:
`var a=localStorage.newTabUrl;chrome.tabs.create({url:a?JSON.parse(a):"/index.html"});window.close();`
* set Chrome's new tab url to `chrome-extension://***/pages/newtab.html`
As a result, `Ctrl+T` will open a tab which opens another "normal" tab and closes itself at once, and in the new page we can see, vimium and focus aree both prepared.
* press `t` will open the same url as what `Ctrl+T` does

**Comment by mfarrugi** [comment]
Is this the same issue as being unable to remove the vimium exclusion on `chrome://newtab/*`?

**Comment by smblott-github** [comment]
@mfarrugi.  If you mean "Vimium does not work on `chrome://newtab`", then yes.

(I wrote a tiny extension ([here](https://chrome.google.com/webstore/detail/new-tab-url/njigpponciklokfkoddampoienefegcl)) to get around this, if you're happy with setting your own new-tab page.)

**Comment by mfarrugi** [comment]
Ah. Unfortunately, like nikess, I would've liked it to work with an extension (SpeedDial2), even if browser focus snapped to the omnibar to start.

Thanks though!

**Comment by rrsahuone** [comment]
We record you best movers and packers that transportation cars and other automobiles as well. They guarantee 100 % protection when products are moved to different position. 

<a href="http://www.expert5th.in/packers-and-movers-hyderabad/" title="Packers and Movers in Hyderabad">Packers and Movers in Hyderabad</a>
<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Packers and Movers in Bangalore">Packers and Movers in Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Packers and Movers in Mumbai">Packers and Movers in Mumbai</a>

**Comment by rrsahuone** [comment]
For moving car to distant place, car providers are used. When we are moving automobiles, we create sure there is not even only one scratch was created on automobile during shifting.

<a href="http://www.expert5th.in/packers-and-movers-gurgaon/" title="Packers and Movers in Gurgaon">Packers and Movers in Gurgaon</a>
<a href="http://www.expert5th.in/packers-and-movers-delhi/" title="Packers and Movers in Delhi">Packers and Movers in Delhi</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Packers and Movers in Pune">Packers and Movers in Pune</a>

**Comment by rrsahuone** [comment]
We are using unique customized containers that help you in moving your automobile safely. We provide you with most expert solutions and estimates from best companies to help you. 

<a href="http://www.expert5th.in/packers-and-movers-chennai/" title="Packers and Movers in Chennai">Packers and Movers in Chennai</a>
<a href="http://www.expert5th.in/packers-and-movers-noida/" title="Packers and Movers in Noida">Packers and Movers in Noida</a>
<a href="http://www.expert5th.in/packers-and-movers-navimumbai/" title="Packers and Movers in Navi Mumbai">Packers and Movers in Navi Mumbai</a>
<a href="http://www.expert5th.in/packers-and-movers-thane/" title="Packers and Movers in Thane">Packers and Movers in Thane</a>
<a href="http://www.expert5th.in/packers-and-movers-ghaziabad/" title="Packers and Movers in Ghaziabad">Packers and Movers in Ghaziabad</a>
<a href="http://www.expert5th.in/packers-and-movers-faridabad/" title="Packers and Movers in Faridabad">Packers and Movers in Faridabad</a>

**Comment by rrsahuone** [comment]
Our detailed movers & packers are extremely effective and can be reliable blindly. They usually seek the services of group of most knowledgeable individuals having outstanding knowledge of transportation department. They are investing fun in moving and shifting either it is little or big.

<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Packers and Movers Bangalore">Packers and Movers Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Packers and Movers Pune">Packers and Movers Pune</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Packers and Movers Mumbai">Packers and Movers Mumbai</a>

**Comment by sunnijie** [comment]
To use vimium commands after creating a new tab with `Ctrl-T`, just press `<tab>` when Chrome focuses the omnibox.

**Comment by Chekote** [comment]
I have this problem with Chrome 59.0.3071.86. It doesn't matter what is focused. If you are on the new tab page, vimium doesn't work at all. You have to navigate to a real webpage for it to work.

**Comment by danielbaniel** [comment]
I have the same issue described by Chekote.
Version 61.0.3163.100 (Official Build) (64-bit)

**Comment by adam-stamand** [comment]
I've got the same issue. After creating a new tab, vimium ceases to do anything at all, regardless of focus. You have to navigate to a different webpage before it begins working. 

Version 62.0.3202.94 (Official Build) Built on Ubuntu , running on Ubuntu 16.04 (64-bit)

**Comment by adam-stamand** [comment]
I should have searched more carefully! The issue is addressed here #2845

**Comment by bschnabel** [comment]
for anyone else searching for a solution: https://github.com/philc/vimium/issues/531#issuecomment-2995207228 might help.
